pilot/cmd/pilot-agent/status/server_test.go

```diff
@@ -255,57 +269,93 @@ func TestAppProbe(t *testing.T) {
-"/app-health/hello-world/livez": {"httpGet":{"port": %v}}}`, appPort, appPort),
-	})
-	if err != nil {
-		t.Errorf("failed to create status server %v", err)
-		return
-	}
-	go server.Run(context.Background())
-
-	var statusPort uint16
-	for statusPort == 0 {
-		server.mutex.RLock()
-		statusPort = server.statusPort
-		server.mutex.RUnlock()
+	simpleConfig := KubeAppProbers{
+		"/app-health/hello-world/readyz": &Prober{
+			HTTPGet: &v1.HTTPGetAction{
+				Path: "/hello/sunnyvale",
+				Port: intstr.IntOrString{IntVal: int32(appPort)},
+			},
+		},
+		"/app-health/hello-world/livez": &Prober{
+			HTTPGet: &v1.HTTPGetAction{
+				Port: intstr.IntOrString{IntVal: int32(appPort)},
+			},
+		},
 	}
-
-	t.Logf("status server starts at port %v, app starts at port %v", statusPort, appPort)
+	_ = simpleConfig
 	testCases := []struct {
 		probePath  string
+		config     KubeAppProbers
 		statusCode int
 	}{
 		{
-			probePath:  fmt.Sprintf(":%v/bad-path-should-be-404", statusPort),
+			probePath:  "bad-path-should-be-404",
+			config:     simpleConfig,
 			statusCode: http.StatusNotFound,
 		},
 		{
-			probePath:  fmt.Sprintf(":%v/app-health/hello-world/readyz", statusPort),
+			probePath:  "app-health/hello-world/readyz",
+			config:     simpleConfig,
 			statusCode: http.StatusOK,
 		},
 		{
-			probePath:  fmt.Sprintf(":%v/app-health/hello-world/livez", statusPort),
+			probePath:  "app-health/hello-world/livez",
+			config:     simpleConfig,
+			statusCode: http.StatusOK,
+		},
+		{
+			probePath: "app-health/header/readyz",
+			config: KubeAppProbers{
+				"/app-health/header/readyz": &Prober{
+					HTTPGet: &v1.HTTPGetAction{
+						Port: intstr.IntOrString{IntVal: int32(appPort)},
+						Path: "/header",
+						HTTPHeaders: []v1.HTTPHeader{
+							{"Host", "host"},
+							{"Some-Header", "some-value"},
+						},
+					},
+				},
+			},
 			statusCode: http.StatusOK,
 		},
 	}
 	for _, tc := range testCases {
-		client := http.Client{}
-		req, err := http.NewRequest("GET", fmt.Sprintf("http://localhost%s", tc.probePath), nil)
-		if err != nil {
-			t.Errorf("[%v] failed to create request", tc.probePath)
-		}
-		resp, err := client.Do(req)
-		if err != nil {
-			t.Fatal("request failed")
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode != tc.statusCode {
-			t.Errorf("[%v] unexpected status code, want = %v, got = %v", tc.probePath, tc.statusCode, resp.StatusCode)
-		}
+		t.Run(tc.probePath, func(t *testing.T) {
+			// Starts the pilot agent status server.
+			server, err := NewServer(Config{
+				StatusPort: 0,
+			})
+			if err != nil {
+				t.Fatalf("failed to create status server %v", err)
+			}
+			server.appKubeProbers = tc.config
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			_ = cancel
+			go server.Run(ctx)
+
+			var statusPort uint16
+			for statusPort == 0 {
+				server.mutex.RLock()
+				statusPort = server.statusPort
+				server.mutex.RUnlock()
+			}
```

Comment:
This is not a good way to wait for server up. It is dead loop.

Reply:
I am just moving the code, it already exists and hasn't seemed to cause issues